storage/bucket.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/coreos/mantle/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/mantle/Godeps/_workspace/src/google.golang.org/api/googleapi"
+	"github.com/coreos/mantle/Godeps/_workspace/src/google.golang.org/api/storage/v1"
+)
+
+var (
+	UnknownScheme = errors.New("storage: URL missing gs:// scheme")
+	UnknownBucket = errors.New("storage: URL missing bucket name")
+)
+
+type Bucket struct {
+	service *storage.Service
+	name    string
+	prefix  string
+
+	mu      sync.RWMutex
+	objects map[string]*storage.Object
+
+	// writeAlways enables overwriting of objects that appear up-to-date
+	writeAlways bool
+	// writeDryRun blocks any changes, merely logging them instead
+	writeDryRun bool
+}
+
+func NewBucket(client *http.Client, bucketURL string) (*Bucket, error) {
+	service, err := storage.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedURL, err := url.Parse(bucketURL)
+	if err != nil {
+		return nil, err
+	}
+	if parsedURL.Scheme != "gs" {
+		return nil, UnknownScheme
+	}
+	if parsedURL.Host == "" {
+		return nil, UnknownBucket
+	}
+
+	// unsure if this is behavior is what we want or not
+	if parsedURL.Path != "" && !strings.HasSuffix(parsedURL.Path, "/") {
+		parsedURL.Path += "/"
+	}
+
+	return &Bucket{
+		service: service,
+		name:    parsedURL.Host,
+		prefix:  strings.TrimPrefix(parsedURL.Path, "/"),
+		objects: make(map[string]*storage.Object),
+	}, nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+func (b *Bucket) Prefix() string {
+	return b.prefix
+}
+
+func (b *Bucket) URL() *url.URL {
+	return &url.URL{Scheme: "gs", Host: b.name, Path: b.prefix}
+}
+
+func (b *Bucket) WriteAlways(always bool) {
+	b.writeAlways = always
+}
+
+func (b *Bucket) WriteDryRun(dryrun bool) {
+	b.writeDryRun = dryrun
+}
+
+func (b *Bucket) Object(objName string) *storage.Object {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.objects[objName]
+}
+
+// Check of an object name's directory is the specified prefix.
+func matchPrefix(path, prefix string) bool {
```

Comment:
No, that would match a name that is nested at any level under a directory, this is matching names that are directly under the given prefix, no further subdirs.

That said, I'm going to delete this since I never wound up using the prefix arg to Objects()